src/server/render.ts

```diff
@@ -190,16 +205,41 @@ export async function render<Data>(
 
   let bodyHtml: string | null = null;
 
-  function renderInner(vnode: ComponentChildren): string {
+  const syncPlugins = opts.plugins.filter((p) => p.render);
+
+  function buildRoot(appComponent: VNode) {
+    return h(CSP_CONTEXT.Provider, {
+      value: csp,
+      children: h(HEAD_CONTEXT.Provider, {
+        value: headComponents,
+        children: h(opts.app.default, {
+          params: opts.params as Record<string, string>,
+          url: opts.url,
+          route: opts.route.pattern,
+          data: opts.data,
+          state: opts.state!,
+          Component() {
+            return appComponent;
+          },
+        }),
+      }),
+    });
+  }
+
+  function renderInnerSync(vnode: ComponentChildren): string {
     // deno-lint-ignore no-explicit-any
     let finalAppComp: VNode<any> = vnode as any;
 
+    if (layouts.some((layout) => checkAsyncComponent(layout.default))) {
+      throw new Error(getSyncPluginsMessage(syncPlugins));
+    }
```

Comment:
Is this assertion able to be moved to earlier in the order of execution? Genuine question.

Reply:
Good call out. Just noticed that this check doesn't make sense here it all. We always start rendering by calling into `renderAsync()` and fall back to sync rendering from there. This means we already run into the other check for async routes + layouts and would never reach this here with an async layout anyway.